src/domain/ledger/imbalance-calculator.ts

```diff
@@ -0,0 +1,50 @@
+const MS_PER_HOUR = (60 * 60 * 1000) as MilliSeconds
+const MS_PER_DAY = (24 * MS_PER_HOUR) as MilliSeconds
+
+export const ImbalanceCalculator = ({
+  volumeLightningFn,
+  volumeOnChainFn,
+  sinceDaysAgo,
+}: ImbalanceCalculatorConfig): ImbalanceCalculator => {
+  const since = new Date(new Date().getTime() - sinceDaysAgo * MS_PER_DAY)
+
+  const getImbalance = async ({
+    volumeFn,
+    walletId,
+    since,
+  }: {
+    volumeFn: GetVolumeSinceFn
+    walletId: WalletId
+    since: Date
+  }) => {
+    const volume_ = await volumeFn({
+      walletId,
```

Comment:
I do wonder if the ideal metric to measure and charge based on is the user's wallet-scoped imbalance.

> To accomplish this in a competitive way, a bank must, among other things, maintain suﬀicient onchain and offchain liquidity. We define the loop imbalance as the difference between the bank’s on-chain liquidity and offchain outbound liquidity.

To me the metric it seems you really care about is the bank's overall _loop imbalance_.

As-is, the way you calculate this dynamic fee seems to econimically incentivize all your users to behave in a homogenous/net-balanced way during the rolling 30 day window. It seems you have a diverse set of users who behave differently (ie b2c merchants vs consumers vs b2b service-providers/producers, etc..) Don't you want to incentivize them all balancing out as a whole, not necessarily individually? That seems to be what your system needs to function smoothly/efficiently?

I'm just wondering if a component of this calculation could be how much this specific transaction will impact the bank's overall _loop imbalance_ (if this txn will improve it maybe this component of the fee should be 0 (or even negative in the future), if this txn further imbalances it maybe this additional fee kicks in?) But maybe this kind of data is not accessible here, in this scope or permission level?

Reply:
> To me the metric it seems you really care about is the bank's overall loop imbalance.

that is correct, but at the level of individual users, the approach in this PR currently only is adding fees for user that are "looping out" our swaping out above some threshold.

This is based on empirical evidence (which are highlighted in the paper) that this is the flow that needs higher fees for a more balanced overall bank imbalance.

Reply:
Makes total sense - thanks for the explanation